prusti-viper/src/encoder/counterexample.rs

```diff
@@ -0,0 +1,169 @@
+                //place is a tuple (Name of the variable, Option<Scope>)
+                if let Some(entry_arg) = self.args.get(place) {
+                    let note = format!("counterexample for \"{0}\"\ninitial: {0} <- {1:#?}\nfinal: {0} <- {2:#?}", 
+                        place.0, 
+                        entry_arg,  
+                        entry,
+                    );
+                    prusti_error = prusti_error.add_note(&note, Some(place.1.clone()));
+                } else {
+                    let note = format!("counterexample for \"{0}\"\n{0} <- {1:#?}", place.0, entry);
+                    prusti_error = prusti_error.add_note(&note, Some(place.1.clone()));
+                }
+            }
+            let result_note = format!("result <- {:#?}", self.result);
+            prusti_error = prusti_error.add_note(&result_note, self.result_span.clone());
+        } else {
+            for (place, entry) in &self.args {
+                let note = format!("counterexample for \"{0}\"\n{0} <- {1:#?}", 
+                        place.0,   
+                        entry,
+                    );
+                prusti_error = prusti_error.add_note(&note, Some(place.1.clone()));
+            }
+            // Todo: find span of return type to give this note a span
+            let result_note = format!("result <- {:#?}", self.result);
+            prusti_error = prusti_error.add_note(&result_note, None);
+        }
+        prusti_error
+    }
+}
+
+impl fmt::Display for Counterexample {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Counterexample:\n");
+        write!(f, "initial args:\n");
+        for (place, entry) in &self.args {
+            let s = format!("{} <- {:?}\n", place.0, entry);
+            write!(f, "{}", indent(s));
+        }
+        write!(f, "\nlocal values when failing:\n");
+        for (place, entry) in &self.entries {
+            let s = format!("{} <- {:?}\n", place.0, entry);
+            write!(f, "{}", indent(s));
+        }
+        write!(f, "\nresult <- {:#?}", self.result)
+    }
+}
+
+pub enum Entry {
+    IntEntry { value: i64 },
+    BoolEntry { value: bool },
+    CharEntry { value: char },
+    RefEntry { el: Box<Entry> },
+    Struct {
+        name: String,
+        field_entries: Vec<(String, Entry)>,
+    },
+    Enum {
+        super_name: String,
+        name: String,
+        field_entries: Vec<(String, Entry)>,
+        //note: if fields are not named, their order is important!
+        //that is why no HashMap is used
+    },
+    Tuple {
+        fields: Vec<Entry>,
+    },
+    Unit,
+    UnknownEntry,
+}
+
+impl fmt::Debug for Entry {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Entry::IntEntry { value } => write!(f, "{}", value),
+            Entry::BoolEntry { value } => write!(f, "{}", value),
+            Entry::CharEntry { value } => write!(f, "'{}' ({:x})", value, *value as i32),
```

Comment:
Reading your evaluation section I realised this can be confusing without the standard `0x` prefix added:

```suggestion
            Entry::CharEntry { value } => write!(f, "'{}' (0x{:x})", value, *value as i32),
```